backend/src/controllers/upload.ts

```diff
@@ -0,0 +1,116 @@
+import express from "express";
+import { Storage } from "@google-cloud/storage";
+import stream from "stream";
+import { Readable } from "stream";
+import fullConfig from "../config/config";
+import crypto from "crypto";
+import { Upload } from "../models/upload";
+
+let storage = new Storage();
+
+if (fullConfig.googleSvcAcctKey !== "") {
+  storage = new Storage({keyFilename: fullConfig.googleSvcAcctKey});
+}
+
+const storageBucketName = fullConfig.storageBucketName
+
+class UploadController {
+  async upload(req: express.Request, res: express.Response) {
+    // check first before proceeding
+    if (req.body.originalFilename === null || req.body.originalFilename === "") {
+      const result = {"status": "failure", "message": "Invalid original filename"};
+      return res.end(JSON.stringify(result));
+    }
+
+    const MAX_UPLOAD_MB = 10;
+    if (req.body.base64File.length / 4 * 3 > MAX_UPLOAD_MB * 1024 * 1024) {
+      const result = {"status": "failure", "message": "File size too large"};
+      return res.end(JSON.stringify(result));
+    }
+
+    const bufferStream = new stream.PassThrough();
+    bufferStream.end(Buffer.from(req.body.base64File, 'base64'));
+
+    //Define bucket.
+    const myBucket = storage.bucket(storageBucketName);
+    //Define file & file name.
+    const newFileName = makeRandomName();
+    const file = myBucket.file(newFileName);
+    //Pipe the 'bufferStream' into a 'file.createWriteStream' method.
+    bufferStream.pipe(file.createWriteStream())
+      .on('error', function(err) {
```

Comment:
'finish' event is not emitted when there's an error, so this callback should return an error response as well

Reply:
Good point. Have added error response after error